overlord/snapstate/backend/setup.go

```diff
@@ -325,3 +289,85 @@ func (b Backend) UndoSetupComponent(cpi snap.ContainerPlaceInfo, installRecord *
 func (b Backend) RemoveSnapInhibitLock(instanceName string) error {
 	return runinhibit.RemoveLockFile(instanceName)
 }
+
+// SetupKernelModulesComponents changes kernel-modules configuration by adding
+// compsToInstall. The components currently active are currentComps, while
+// ksnapName and ksnapRev identify the currently active kernel.
+func (b Backend) SetupKernelModulesComponents(compsToInstall, currentComps []*snap.ComponentSideInfo, ksnapName string, ksnapRev snap.Revision, meter progress.Meter) (err error) {
+	sysd := newSystemd(b.preseed, meter)
+
+	// newActiveComps will contain the new revisions of components, taken from compsToInstall
+	newActiveComps, _ := mergeCompSideInfosUpdatingRev(currentComps, compsToInstall)
+
+	return moveKModsComponentsState(
+		currentComps, newActiveComps, ksnapName, ksnapRev, sysd,
+		"after failure to set-up kernel modules components")
+}
+
+// RemoveKernelModulesComponentsSetup changes kernel-modules configuration by
+// removing compsToRemove and making the final state consider only finalComps.
+func (b Backend) RemoveKernelModulesComponentsSetup(compsToRemove, finalComps []*snap.ComponentSideInfo, ksnapName string, ksnapRev snap.Revision, meter progress.Meter) (err error) {
+	sysd := newSystemd(b.preseed, meter)
+
+	// currentActiveComps will contain the current revision, taken from compsToRemove
+	currentActiveComps, _ := mergeCompSideInfosUpdatingRev(finalComps, compsToRemove)
+
+	return moveKModsComponentsState(
+		currentActiveComps, finalComps, ksnapName, ksnapRev, sysd,
+		"after failure to remove set-up of kernel modules components")
+}
+
+// mergeCompSideInfosUpdatingRev returns a merged list from two lists of
+// ComponentSideInfo, and a list with the intersection, using the criteria of
+// the elements having the same ComponentRef. The rest of the data for an
+// element will come from comps2 if ComponentRef is the same in comps1 and
+// comps2, that is, the revision is updated in that case.
+func mergeCompSideInfosUpdatingRev(comps1, comps2 []*snap.ComponentSideInfo) (merged, instersect []*snap.ComponentSideInfo) {
```

Comment:
btw. interesect isn't used by any of the callers, is there a PR in queue which will make use of it? if not, I'd drop it for now.

Reply:
Yeah, it is not used now, I've removed in the end.